cluster-autoscaler/proposals/min_at_zero_gcp.md

```diff
@@ -0,0 +1,176 @@
+      value:
+        ENABLE_NODE_PROBLEM_DETECTOR: 'daemonset'
+        ENABLE_RESCHEDULER: 'true'
+        LOGGING_DESTINATION: 'gcp'
+        ELASTICSEARCH_LOGGING_REPLICAS: '1'
+        ENABLE_CLUSTER_DNS: 'true'
+        ENABLE_CLUSTER_REGISTRY: 'false'
+        CLUSTER_REGISTRY_DISK: 'kubernetes-kube-system-kube-registry'
+        CLUSTER_REGISTRY_DISK_SIZE: '200GB'
+        DNS_SERVER_IP: '10.0.0.10'
+        DNS_DOMAIN: 'cluster.local'
+        ENABLE_DNS_HORIZONTAL_AUTOSCALER: 'true'
+        [...]
+```
+
+Node type and kube-env. These two fields are very useful to estimate the node shape.
+
+### [P1A] - Node capacity
+To estimate the node capacity we need to decode n1-standard-2 or any other string mentioned in the machineType. The format of the strings is well documented
+here: https://cloud.google.com/compute/docs/machine-types
+
+However, except for the earliest prototype, we probably don’t want to manually hardcode all of the values available there. Luckily, the GCP platform provides an api to get the machine description in terms of memory and cpu:
+
+```json
+{
+ "kind": "compute#machineTypeList",
+ "id": "projects/mwielgus-proj/zones/us-central1-b/machineTypes",
+ "items": [
+  {
+   "kind": "compute#machineType",
+   "id": "2000",
+   "creationTimestamp": "2015-01-16T09:25:43.315-08:00",
+   "name": "g1-small",
+   "description": "1 vCPU (shared physical core) and 1.7 GB RAM",
+   "guestCpus": 1,
+   "memoryMb": 1740,
+   "maximumPersistentDisks": 16,
+   "maximumPersistentDisksSizeGb": "3072",
+   "zone": "us-central1-b",
+   "selfLink": "https://www.googleapis.com/compute/v1/projects/mwielgus-proj/zones/us-central1-b/machineTypes/g1-small",
+   "isSharedCpu": true
+  },
+  {
+   "kind": "compute#machineType",
+   "id": "4016",
+   "creationTimestamp": "2015-01-16T09:25:43.316-08:00",
+   "name": "n1-highcpu-16",
+   "description": "16 vCPUs, 14.4 GB RAM",
+   "guestCpus": 16,
+   "memoryMb": 14746,
+   "maximumPersistentDisks": 128,
+   "maximumPersistentDisksSizeGb": "65536",
+   "zone": "us-central1-b",
+   "selfLink": "https://www.googleapis.com/compute/v1/projects/mwielgus-proj/zones/us-central1-b/machineTypes/n1-highcpu-16",
+   "isSharedCpu": false
+  },
+```
+
+So all standard type sizes can be obtained by API. On GCP there is a possibility of a custom machine type. This custom machine type has the name formatted as:
+
+```
+custom-<cpu_count>-<memory_in_mb>
+```
+
+So it also quite easy to get all of capacity information from it.
+
+### [P1B] - Node allocatable
+
+In GKE 1.5.6 allocatable for new nodes is equal to capacity. For simplicity we could assume that the new node will have the 90% (or -0.1cpu/-200mb) of capacity. Being wrong or underestimating here is not fatal, most users will probably be OK with this. Once some nodes are present we will have more precise estimates. The worst thing that can happen is that the scale up may not be triggered if the request is exactly at the node capacity - system pods.
```

Comment:
That is not true on non-GKE cluster on GKE. We could at least mention that fact and put some sort of TODO to revisit this.